Kernel/Devices/Storage/USB/BulkSCSIInterface.cpp

```diff
@@ -16,6 +16,109 @@ BulkSCSIInterface::BulkSCSIInterface(LUNAddress logical_unit_number_address, u32
+            // The available pages are (supposedly) sorted in ascending order
+            // so we can break here early
+            break;
+        }
+    }
+
+    if (!found_block_limits) {
+        dmesgln("SCSI/BBB: Device does not support block limits page");
+        // This is not an error, we just won't be able to optimize our transfers
+        return {};
+    }
+
+    inquiry_command.page_code = SCSI::VitalProductDataPageCode::BlockLimits;
+    SCSI::BlockLimitsPage block_limits_page {};
+    inquiry_command.allocation_length = sizeof(SCSI::BlockLimitsPage);
+    // FIXME: Technically we should first ask for the supported vital product data pages
+    status = TRY(send_scsi_command<SCSIDataDirection::DataToInitiator>(*m_out_pipe, *m_in_pipe, inquiry_command, &block_limits_page, sizeof(SCSI::BlockLimitsPage)));
+    if (status.status != CSWStatus::Passed) {
+        dbgln("SCSI/BBB: Inquiry failed to inquire block limits with code {}", to_underlying(status.status));
+        // FIXME: Maybe request sense here
+    }
+
+    if (block_limits_page.page_code != SCSI::VitalProductDataPageCode::BlockLimits) {
+        dmesgln("SCSI/BBB: Returned wrong page code for block limits {:#02x}", to_underlying(block_limits_page.page_code));
+        return EIO;
+    }
+
+    if (block_limits_page.page_length != sizeof(SCSI::BlockLimitsPage) - 4) {
+        dmesgln("SCSI/BBB: Returned wrong page length for block limits {}", block_limits_page.page_length);
+        return EIO;
+    }
+
+    if (block_limits_page.maximum_transfer_length != 0)
+        m_maximum_transfer_length = block_limits_page.maximum_transfer_length;
+    if (block_limits_page.optimal_transfer_length != 0)
+        m_optimal_transfer_length = block_limits_page.optimal_transfer_length;
+    if (block_limits_page.optimal_transfer_length_granularity != 0)
+        m_optimal_transfer_length_granularity = block_limits_page.optimal_transfer_length_granularity;
+
+    dbgln("SCSI/BBB: Maximum transfer length: {}", m_maximum_transfer_length);
+    dbgln("SCSI/BBB: Optimal transfer length: {}", m_optimal_transfer_length);
+    dbgln("SCSI/BBB: Optimal transfer length granularity: {}", m_optimal_transfer_length_granularity);
+
+    return {};
+}
+
+u32 BulkSCSIInterface::optimal_block_count(u32 blocks)
+{
+    if (m_maximum_transfer_length.has_value() && blocks > m_maximum_transfer_length.value())
+        return m_maximum_transfer_length.value();
+    // FIXME: Whats the best way of doing this?
+    //        What takes precedence, granularity or length?
+    //        What is usually bigger?
+    if (m_optimal_transfer_length.has_value() && blocks > m_optimal_transfer_length.value())
+        return m_optimal_transfer_length.value();
+
+    if (!m_optimal_transfer_length_granularity.has_value())
+        return blocks;
+
+    return blocks - (blocks % m_optimal_transfer_length_granularity.value());
```

Comment:
This doesn't work when `blocks < m_optimal_transfer_length_granularity.value()`.
Also, isn't that just reimplementing `align_down_to`?

Reply:
Essentially, but `align_down_to` only works with powers of 2, which should be a fair assumption, but the spec does not say anything on that

Reply:
Ah right, we don't seem to have a function for that. But the code is simple enough tbh.